File: exercises/Exercise.1-DerivedDataTypes/.soln/f90/ddt.4.struct/mpi_ddt_struct.f90
program struct
    include 'mpif.h'

    integer NELEM
    parameter(NELEM=25)
    integer numtasks, rank, source, dest, tag, i,  ierr
    integer stat(MPI_STATUS_SIZE)

    type Particle
      sequence
        real*4 x, y, z, velocity
        integer n, type
    end type Particle

    type (Particle) p(NELEM), particles(NELEM)
    integer particletype, oldtypes(0:1)   ! required variables
    integer blockcounts(0:1), offsets(0:1), extent
    tag = 1

    call MPI_INIT(ierr)
    call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

    ! =====================================================================
    ! Step 1. Create an MPI Struct Type
    !    Summary:
    !      Make a new struct derived datatype.
    !
    !    Function Call:
    !       MPI_TYPE_STRUCT(INTEGER COUNT, 
    !                       INTEGER(*) ARRAY_OF_BLOCKLENGTHS,
    !                       INTEGER(*) ARRAY_OF_DISPLACEMENTS,
    !                       INTEGER(*) ARRAY_OF_TYPES,
    !                       INTEGER NEWTYPE,
    !                       INTEGER IERROR)
    !
    !   Input Parameters:
    !       COUNT
    !          number of blocks (integer) -- also number of entries in arrays array_of_types , array_of_displacements and array_of_blocklengths
    !       ARRAY_OF_BLOCKLENGTHS
    !          number of elements in each block (array)
    !       ARRAY_OF_DISPLACEMENTS
    !          byte displacement of each block (array)
    !       ARRAY_OF_TYPES
    !          type of elements in each block (array of handles to datatype objects)
    !
    !   Output Parameters:
    !      NEWTYPE
    !          new datatype (handle)
    !      IERROR
    !          Fortran only: Error status (integer).
    !
    ! TODO: setup description of the 4 MPI_REAL fields x, y, z, velocity
    offsets(0) = 0 ! TODO
    oldtypes(0) = MPI_REAL ! TODO
    blockcounts(0) = 4 ! TODO

    ! TODO: setup description of the 2 MPI_INTEGER fields n, type 
    !       We first figure offset by getting size of MPI_REAL
    call MPI_TYPE_EXTENT(MPI_REAL, extent, ierr)
    offsets(1) = 4 * extent ! TODO; HINT: how many 'extent's do we need?
    oldtypes(1) = MPI_INTEGER  ! TODO 
    blockcounts(1) = 2  ! TODO

    ! TODO: create the struct data type
    call MPI_TYPE_STRUCT(2, blockcounts, offsets, oldtypes, particletype, ierr)
    ! TODO: commit the new derived datatype 
    call MPI_TYPE_COMMIT(particletype, ierr)
    ! =====================================================================

    ! task 0 initializes the particle array and then sends it to each task
    if (rank .eq. 0) then
        do i=1, NELEM
            particles(i) = Particle ( 1.0*i, -1.0*i, 1.0*i, 0.25, i, mod(i,2) )
        end do

        do i=0, numtasks-1
            call MPI_SEND(particles, NELEM, particletype, i, tag, &
                          MPI_COMM_WORLD, ierr)
        end do
    endif

    ! all tasks receive particletype data
    source = 0
    call MPI_RECV(p, NELEM, particletype, source, tag, &
                  MPI_COMM_WORLD, stat, ierr)

    print *, 'rank= ',rank,' p(3)= ',p(3)

    ! free datatype when done using it
    call MPI_TYPE_FREE(particletype, ierr)
    call MPI_FINALIZE(ierr)
end program

File: exercises/Exercise.1-DerivedDataTypes/c/ddt.4.struct/mpi_ddt_struct.c
#include "mpi.h"
#include <stdio.h>
#define NELEM 25

int main(int argc, char *argv[])  {
    int numtasks, rank, source=0, dest, tag=1, i;

    typedef struct {
        float x, y, z;
        float velocity;
        int  n, type;
    } Particle;

    Particle p[NELEM], particles[NELEM];
    MPI_Datatype particletype, oldtypes[2];   // required variables
    int blockcounts[2];

    // MPI_Aint type used to be consistent with syntax of
    // MPI_Type_extent routine
    MPI_Aint offsets[2], extent;
    MPI_Status stat;

    MPI_Init(&argc,&argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &numtasks);

    /* ===================================================================== */
    /* Step 1. Create an MPI Struct Type
     *    Summary:
     *      Make a new struct derived datatype.
     *
     *    Function Call:
     *       int MPI_Type_struct(int count,
     *                           const int *array_of_blocklengths,
     *                           const MPI_Aint *array_of_displacements,
     *                           const MPI_Datatype *array_of_types,
     *                           MPI_Datatype *newtype);
     *
     *   Input Parameters:
     *       count
     *          number of blocks (integer) -- also number of entries in arrays array_of_types , array_of_displacements and array_of_blocklengths
     *       array_of_blocklengths
     *          number of elements in each block (array)
     *       array_of_displacements
     *          byte displacement of each block (array)
     *       array_of_types
     *          type of elements in each block (array of handles to datatype objects)
     *
     *   Output Parameters:
     *      newtype
     *          new datatype (handle)
     */
    // TODO: setup description of the 4 MPI_FLOAT fields x, y, z, velocity
    offsets[0] = //TODO;
    oldtypes[0] = //TODO;
    blockcounts[0] = //TODO;

    // TODO: setup description of the 2 MPI_INT fields n, type.
    //      We first figure offset by getting size of MPI_FLOAT
    MPI_Type_extent(MPI_FLOAT, &extent);
    offsets[1] = //TODO; HINT: how many 'extent's do we need?
    oldtypes[1] = //TODO;
    blockcounts[1] = //TODO;
    
    // TODO: create the struct data type

    // TODO: commit the new derived datatype 

    /* ===================================================================== */

    // task 0 initializes the particle array and then sends it to each task
    if (rank == 0) {
        for (i=0; i<NELEM; i++) {
            particles[i].x = i * 1.0;
            particles[i].y = i * -1.0;
            particles[i].z = i * 1.0; 
            particles[i].velocity = 0.25;
            particles[i].n = i;
            particles[i].type = i % 2; 
        }
        for (i=0; i<numtasks; i++) {
            MPI_Send(particles, NELEM, particletype, i, tag, MPI_COMM_WORLD);
        }
    }

    // all tasks receive particletype data
    MPI_Recv(p, NELEM, particletype, source, tag, MPI_COMM_WORLD, &stat);

    printf("rank= %d   %3.2f %3.2f %3.2f %3.2f %d %d\n", rank,p[3].x,
           p[3].y,p[3].z,p[3].velocity,p[3].n,p[3].type);

    // free datatype when done using it
    MPI_Type_free(&particletype);
    MPI_Finalize();
    return 0;
}
